TOOLS/lua/autocrop.lua

```diff
@@ -1,114 +1,258 @@
-                                     crop_label,
-                                     cropdetect_metadata["lavfi.cropdetect.w"],
-                                     cropdetect_metadata["lavfi.cropdetect.h"],
-                                     cropdetect_metadata["lavfi.cropdetect.x"],
-                                     cropdetect_metadata["lavfi.cropdetect.y"]))
-        else
-            mp.msg.error(
-                "Got empty crop data. You might need to increase detect_seconds."
-            )
-        end
+        meta = {
+            w = cropdetect_metadata["lavfi.cropdetect.w"],
+            h = cropdetect_metadata["lavfi.cropdetect.h"],
+            x = cropdetect_metadata["lavfi.cropdetect.x"],
+            y = cropdetect_metadata["lavfi.cropdetect.y"],
+        }
     else
-        mp.msg.error(
-            "No crop data. Was the cropdetect filter successfully inserted?"
-        )
-        mp.msg.error(
-            "Does your version of ffmpeg/libav support AVFrame metadata?"
+        mp.msg.error("No crop data.")
+        mp.msg.info("Was the cropdetect filter successfully inserted?")
+        mp.msg.info("Does your version of ffmpeg/libav support AVFrame metadata?")
+        return
+    end
+
+
+    -- Verify that the metadata meets the requirements and convert it.
+    if meta.w and meta.h and meta.x and meta.y then
+        meta = {
+            w = tonumber(meta.w),
+            h = tonumber(meta.h),
+            x = tonumber(meta.x),
+            y = tonumber(meta.y),
+            min_w = mp.get_property_native("width") * options.detect_min_ratio,
+            min_h = mp.get_property_native("height") * options.detect_min_ratio,
+            max_w = mp.get_property_native("width"),
+            max_h = mp.get_property_native("height")
+        }
+    else
+        mp.msg.error("Got empty crop data.")
+        mp.msg.info("You might need to increase detect_seconds.")
+        return
+    end
+
+
+    apply_crop(meta)
+end
+
+
+function apply_crop(meta)
+
+    -- Verify if it is necessary to crop.
+    local is_effective = meta.x > 0 or meta.y > 0
+        or meta.w < meta.max_w or meta.h < meta.max_h
+
+    if not is_effective then
+        mp.msg.info("No area detected for cropping.")
+        return
+    end
+
+
+    -- Verify if it is over cropped.
+    local is_not_excessive = meta.w >= meta.min_w and meta.h >= meta.min_h
+
+    if not is_not_excessive then
+        mp.msg.info("The area to be cropped is too large.")
+        mp.msg.info("You might need to decrease detect_min_ratio.")
+        return
+    end
+
+
+    -- Remove existing crop.
+    remove_filter(labels.crop)
+
+
+    -- Apply crop.
+    mp.command(
+        string.format("vf pre @%s:lavfi-crop=w=%s:h=%s:x=%s:y=%s",
+            labels.crop, meta.w, meta.h, meta.x, meta.y
         )
+    )
+end
+
+
+function on_start()
+
+    -- Remove all existing filters.
+    remove_filter_all()
```

Comment:
If there is already an auto_delay detect_crop timer (for example if the playlist was advanced before it could fire) it should be stopped. Likewise for mp.add_timeout(options.detect_seconds, detect_end)